src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObjectNative.cs

```diff
@@ -309,8 +330,49 @@ internal override uint SetConnectionBufferSize(ref uint unsignedPacketSize)
         internal override uint GenerateSspiClientContext(byte[] receivedBuff, uint receivedLength, ref byte[] sendBuff, ref uint sendLength, byte[] _sniSpnBuffer)
             => SNINativeMethodWrapper.SNISecGenClientContext(Handle, receivedBuff, receivedLength, sendBuff, ref sendLength, _sniSpnBuffer);
 
-        internal override uint WaitForSSLHandShakeToComplete()
-            => SNINativeMethodWrapper.SNIWaitForSSLHandshakeToComplete(Handle, GetTimeoutRemaining());
+        internal override uint WaitForSSLHandShakeToComplete(out int protocolVersion)
+        {
+            uint returnValue = SNINativeMethodWrapper.SNIWaitForSSLHandshakeToComplete(Handle, GetTimeoutRemaining(), out uint nativeProtocolVersion);
+            var nativeProtocol = (NativeProtocols)nativeProtocolVersion;
+
+            /* The SslProtocols.Tls13 is supported by netcoreapp3.1 and later
+             * This driver does not support this version yet!
+            if (nativeProtocol.HasFlag(NativeProtocols.SP_PROT_TLS1_3_CLIENT) || nativeProtocol.HasFlag(NativeProtocols.SP_PROT_TLS1_3_SERVER))
+            {
+                protocolVersion = (int)SslProtocols.Tls13;
+            }*/
+            if (nativeProtocol.HasFlag(NativeProtocols.SP_PROT_TLS1_2_CLIENT) || nativeProtocol.HasFlag(NativeProtocols.SP_PROT_TLS1_2_SERVER))
+            {
+                protocolVersion = (int)SslProtocols.Tls12;
+            }
+            else if (nativeProtocol.HasFlag(NativeProtocols.SP_PROT_TLS1_1_CLIENT) || nativeProtocol.HasFlag(NativeProtocols.SP_PROT_TLS1_1_SERVER))
+            {
+                protocolVersion = (int)SslProtocols.Tls11;
+            }
+            else if (nativeProtocol.HasFlag(NativeProtocols.SP_PROT_TLS1_0_CLIENT) || nativeProtocol.HasFlag(NativeProtocols.SP_PROT_TLS1_0_SERVER))
+            {
+                protocolVersion = (int)SslProtocols.Tls;
+            }
+            else if (nativeProtocol.HasFlag(NativeProtocols.SP_PROT_SSL3_CLIENT) || nativeProtocol.HasFlag(NativeProtocols.SP_PROT_SSL3_SERVER))
+            {
+#pragma warning disable CS0618 // Type or member is obsolete : SSL is depricated
+                protocolVersion = (int)SslProtocols.Ssl3;
+            }
+            else if (nativeProtocol.HasFlag(NativeProtocols.SP_PROT_SSL2_CLIENT) || nativeProtocol.HasFlag(NativeProtocols.SP_PROT_SSL2_SERVER))
+            {
+                protocolVersion = (int)SslProtocols.Ssl2;
+#pragma warning restore CS0618 // Type or member is obsolete : SSL is depricated
+            }
+            else if(nativeProtocol.HasFlag(NativeProtocols.SP_PROT_NONE))
+            {
+                protocolVersion = (int)SslProtocols.None;
```

Comment:
Will this trigger warning CS0618 since SSL is deprecated?

And line 364 will not be reached when we hit line 358.

Reply:
Yes, this warning triggers for obsolete members. You can try it by removing this line.
[#pragma warning](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/preprocessor-directives/preprocessor-pragma-warning) is a preprocessor directive and you can use it everywhere in the code.

Reply:
> And line 364 will not be reached when we hit line 358.

Yeah, I just realize the preprocessor won't be impacted by common coding logic. Just ignore this comment. 

> Will this trigger warning CS0618 since SSL is deprecated?

This is for `protocolVersion = (int)SslProtocols.None;` which is not covered by `#pragma warning disable CS0618 `.

Reply:
The `SslProtocols.None` does not trigger warning CS0618. It just lets the operating system to choose the best protocol by itself when you want to establish a secure connection([MS document](https://docs.microsoft.com/en-us/dotnet/api/system.security.authentication.sslprotocols?f1url=https%3A%2F%2Fmsdn.microsoft.com%2Fquery%2Fdev16.query%3FappId%3DDev16IDEF1%26l%3DEN-US%26k%3Dk(System.Security.Authentication.SslProtocols.None);k(SolutionItemsProject);k(DevLang-csharp)%26rd%3Dtrue&view=netcore-3.1)).
I hope I could answer your question.